### PR TITLE
ARROW-8950: [C++] Avoid HEAD when possible in S3 filesystem

### PR DESCRIPTION
**Pull request by pitrou**
Add FileSystem::OpenInput{Stream,File} overrides that accept a FileInfo parameter.
This can be used to optimize file opening when it the file size and existence is already known.  Concretely, avoids a HEAD request in S3.

### CODE REVIEW

**Comment by pitrou** [comment]
@bkietz Would like your input on the dataset changes.

**Comment by pitrou** [comment]
Note to self: instead of basing the default `OpenInputStream(const FileInfo&)` on `OpenInputStream(const std::string&)`, it would be more logical to do the reverse, so that no subclass needs to override both.

File: cpp/src/arrow/dataset/discovery.cc [hunk 1]
```diff
@@ -156,27 +177,26 @@ Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
 
   ARROW_ASSIGN_OR_RAISE(auto files, filesystem->GetFileInfo(selector));
 
-  std::vector<std::string> paths;
+  // XXX We may avoid copying around vectors of FileInfo by filtering in place.
+  std::vector<fs::FileInfo> filtered_files;
```
Comment (bkietz) [inline cpp/src/arrow/dataset/discovery.cc:181]:
```suggestion
  auto files_end = std::remove_if(files.begin(), files.end(), [&](const FileInfo& info) {
    if (!info.IsFile()) {
      return true;
    }

    if (StartsWithAnyOf(info.path(), options.selector_ignore_prefixes)) {
      return true;
    }
    
    return false;
  });
  files.erase(files_end, files.end());
```

**Review by bkietz — APPROVED** [review]
The dataset changes look fine to me

File: cpp/src/arrow/dataset/file_parquet.cc [hunk 2]
```diff
@@ -571,6 +571,8 @@ static inline Result<std::string> FileFromRowGroup(
       }
     }
 
+    // TODO Is it possible to infer the file size and return a populated FileInfo?
+    // This could avoid some spurious HEAD requests on S3 (ARROW-8950)
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet.cc:575]: https://issues.apache.org/jira/browse/ARROW-9227

File: cpp/src/arrow/dataset/discovery.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/dataset/discovery.cc:181]: Yeah, that's what I thought :-) Thanks for spelling it out, will give it a try.

File: cpp/src/arrow/filesystem/s3fs.cc [hunk 3]
```diff
@@ -397,9 +399,17 @@ class ObjectInputFile : public io::RandomAccessFile {
   ObjectInputFile(Aws::S3::S3Client* client, const S3Path& path)
       : client_(client), path_(path) {}
 
+  ObjectInputFile(Aws::S3::S3Client* client, const S3Path& path, int64_t size)
+      : client_(client), path_(path), content_length_(size) {}
+
   Status Init() {
     // Issue a HEAD Object to get the content-length and ensure any
     // errors (e.g. file not found) don't wait until the first Read() call.
```
Comment (rdettai) [inline cpp/src/arrow/filesystem/s3fs.cc:407]: This comments got a little bit out of sync with the code below ;-)

**Comment by rdettai** [comment]
Great! AWS is going to be surprised to see its worldwide S3 HEAD request rate drop by half overnight !

**Pull request closed without merging**